### PR TITLE
drivers/ina220: Changed to comply with RIOT's driver design goals & SAUL adaption

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description
This PR contains some major changes to the INA220 power/current monitor driver

1. Rename to INA2XX, as the same driver works for (at least) the INA220 and the INA219, most likely for a bunch of other I2C power/current monitors of the same family
2. Brought the API in compliance with RIOT's driver API goals
    - Use standard RIOT driver initialization/configuration via `ina2xx_params_t` instead of custom API
    - Perform a device reset and connectivity test upon initialization as required by the design goals stated in [RIOT's driver guide][guide]
    - Convert the results directly to reasonable physical units  as required by the design goals stated in [RIOT's driver guide][guide]
3. Integrated the driver into SAUL

[guide]: https://github.com/RIOT-OS/RIOT/wiki/Guide:-Writing-a-device-driver-in-RIOT

### Testing procedure

You'll need an INA219 or INA220 device, ideally hooked up at I2C interface 0 with the address inputs set to zero and a shunt resistor of 100 mΩ, as then the default configuration will work out of the box. (This default config will work for the popular INA219 break out boards, e.g. the ones by Adafruit or any clone.)

The default config is optimized to measure power/current for high-power USB devices (500mA @ 5V) with a comfortable margin (up to 655mA and up to 16V). This would also work for 12V power supplies as are popular e.g. with the Arduino boards.

1. Flash `examples/saul` with `USEMODULE=ina2xx make BOARD=<foo> flash` and interact with the saul shell command
2. Flash `tests/driver_ina2xx` and check the output. The values shown should be plausible

### Issues/PRs references
None

### CODE REVIEW

**Comment by maribu** [comment]
It would be nice if @gebart could review this PR, or - if he is busy - at least check the API changes. I do believe that the API changes are necessary to bring the driver in compliance with the RIOT's device driver goals and reasonable, but this change will break any existing code using the current driver :-(

**Comment by maribu** [comment]
Please do not start reviewing right now, I'll fix the doc issue detected by travis and squash right away. Thx :-)

**Comment by maribu** [comment]
Travis is happy now, I will  refrain from force-pushing now :-)

File: tests/driver_ina2xx/README.md [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # About
-This is a manual test application for the INA220 current and power monitor driver.
+This is a manual test application for the INA210/INA220 current and power
```
Comment (smlng) [inline tests/driver_ina2xx/README.md]: nit-picky: the test app uses `ina219` according to the Makefile

Comment (maribu) [reply tests/driver_ina2xx/README.md]: You're right!

Comment (maribu) [reply tests/driver_ina2xx/README.md]: Also it is completely outdated :-( Good, that I look at it again

**Comment by maribu** [comment]
The first fixup fixes the documentation in the test, the second fixup increases the width in a column of the output of the test. (The shunt voltage could easily require more digit than space was provided prior to the fixup.)

**Comment by maribu** [comment]
This should be rebased on top of #11692 to avoid code duplication

**Comment by maribu** [comment]
OK, I rebased the PR on top of current master, as `fmt_table` is now merged. With that, I could reuse the table printing code in the test application and simplify the PR.

**Comment by maribu** [comment]
Rebased to resolve merge conflics.

@smlng, @MrKevinWeiss: You don't happen to use INA219 or INA220 for your power managements?

**Comment by MrKevinWeiss** [comment]
Only a bit of ina226 but not within RIOT.

**Comment by maribu** [comment]
I bet they work with this driver as well

File: drivers/include/saul.h [hunk 2]
```diff
@@ -105,6 +105,7 @@ enum {
     SAUL_SENSE_CAPACITANCE = 0x97,     /**< sensor: capacitance */
     SAUL_SENSE_VOLTAGE     = 0x98,     /**< sensor: voltage */
     SAUL_SENSE_PH          = 0x99,     /**< sensor: pH */
+    SAUL_SENSE_POWER       = 0x9a,     /**< sensor: (electrical) power */
```
Comment (gschorcht) [inline drivers/include/saul.h]: Is it necessary to specify that electrical power is meant? If not, we could drop it. If yes, I wouldn't place it in braces.

Comment (maribu) [reply drivers/include/saul.h]: You're totally right. The physical unit will be Watt regardless of what power we will speak and having the electrical written there will trigger new entries for each other kind of power, even though there is no reason to differentiate.

**Comment by maribu** [comment]
I'll do a rebase to solve the merge conflict. (No changes in the PR's code now, except for manual merging where needed.)

File: drivers/include/saul.h (hunk 2 above)
Comment (maribu) [reply drivers/include/saul.h]: That sensor type was added to master in the meantime with the same comment. I changed the commit to add this during rebasing to a commit that removes the `(electrical)`.

File: drivers/ina2xx/ina2xx.c [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    uint16_t tmp;
+    int status = ina2xx_read_reg(dev, INA2XX_REG_BUS_VOLTAGE, &tmp);
+    if (status < 0) {
+        return status;
+    }
+
+    /* The voltage given by bits 15-3 in steps of 4 mV
+     * ==> shift 3 to the right and multiply by 4 to get voltage in mV
+     */
+    *voltage = (tmp >> 3) * 4;
```
Comment (benpicco) [inline drivers/ina2xx/ina2xx.c]: For a moment I thought the compiler could optimize it wrong, but probably not.

```suggestion
    *voltage = (tmp & ~0x7) >> 1;
```

will still generate [slightly smaller](https://godbolt.org/z/Bh9vXg) code, at least on AVR.

File: examples/default/Makefile [hunk 4]
```diff
@@ -33,6 +33,7 @@ USEMODULE += shell_commands
 USEMODULE += ps
 # include and auto-initialize all available sensors
 USEMODULE += saul_default
+USEMODULE += ina219
```
Comment (benpicco) [inline examples/default/Makefile]: huh?
I don't think this belongs here.

File: sys/auto_init/saul/auto_init_ina2xx.c [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+#include "assert.h"
+#include "log.h"
+#include "saul_reg.h"
+#include "ina2xx_params.h"
+#include "ina2xx.h"
+
+/**
+ * @brief   Define the number of configured sensors
+ */
+#define INA2XX_NUM     (sizeof(ina2xx_params) / sizeof(ina2xx_params[0]))
```
Comment (benpicco) [inline sys/auto_init/saul/auto_init_ina2xx.c]:
```suggestion
#define INA2XX_NUM     ARRAY_SIZE(ina2xx_params)
```

File: sys/auto_init/saul/auto_init_ina2xx.c [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * @brief   Memory for the SAUL registry entries
+ */
+static saul_reg_t saul_entries[INA2XX_NUM * 3];
+
+/**
+ * @brief   Define the number of saul info
+ */
+#define INA2XX_INFO_NUM (sizeof(ina2xx_saul_info) / sizeof(ina2xx_saul_info[0]))
```
Comment (benpicco) [inline sys/auto_init/saul/auto_init_ina2xx.c]:
```suggestion
#define INA2XX_INFO_NUM ARRAY_SIZE(ina2xx_saul_info)
```

File: sys/auto_init/saul/auto_init_ina2xx.c [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        saul_entries[i * 3    ].driver = &ina2xx_saul_current_driver;
+        saul_entries[i * 3 + 1].dev = &(ina2xx_devs[i]);
+        saul_entries[i * 3 + 1].name = ina2xx_saul_info[3 * i + 1].name;
+        saul_entries[i * 3 + 1].driver = &ina2xx_saul_power_driver;
+        saul_entries[i * 3 + 2].dev = &(ina2xx_devs[i]);
+        saul_entries[i * 3 + 2].name = ina2xx_saul_info[3 * i + 2].name;
+        saul_entries[i * 3 + 2].driver = &ina2xx_saul_voltage_driver;
+        saul_reg_add(&(saul_entries[i * 3    ]));
+        saul_reg_add(&(saul_entries[i * 3 + 1]));
+        saul_reg_add(&(saul_entries[i * 3 + 2]));
```
Comment (benpicco) [inline sys/auto_init/saul/auto_init_ina2xx.c]: While I appreciate the careful alignment, a
```C
for (int j = 0; j < 3; ++i) {
    saul_entries[i * 3 + j].dev = &(ina2xx_devs[i]);
    …
```
would be the more readable / maintainable option :wink:

File: drivers/ina2xx/ina2xx.c [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    return 0;
+}
+
+/** @brief Write one 16 bit register to a INA2XX device and swap byte order, if necessary. */
+static int ina2xx_write_reg(const ina2xx_t *dev, uint8_t reg, uint16_t in)
+{
+    union {
+        uint8_t c[2];
+        uint16_t u16;
+    } tmp = { .u16 = 0 };
```
Comment (benpicco) [inline drivers/ina2xx/ina2xx.c]: dito

File: drivers/ina2xx/ina2xx.c [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/** @brief Read one 16 bit register from a INA2XX device and swap byte order, if necessary. */
+static int ina2xx_read_reg(const ina2xx_t *dev, uint8_t reg, uint16_t *out)
+{
+    union {
+        uint8_t c[2];
+        uint16_t u16;
+    } tmp = { .u16 = 0 };
```
[the remark this answers is not in the document]
Comment (maribu) [reply drivers/ina2xx/ina2xx.c]: Done

File: drivers/ina2xx/ina2xx.c [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+        return -ENODEV;
+    }
+
+    status = ina2xx_write_reg(dev, INA2XX_REG_CONFIGURATION, params->config);
+    if (status < 0) {
+        return status;
+    }
+
+    /* Set calibration so that current register resolution is 0.01 mA */
+    uint32_t calib = 4096000 / params->rshunt_mohm;
```
[the remark this answers is not in the document]
Comment (maribu) [reply drivers/ina2xx/ina2xx.c]: Done

File: drivers/ina2xx/ina2xx.c [hunk 11]
```diff
@@ -106,8 +116,12 @@ int ina2xx_init(ina2xx_t *dev, const ina2xx_params_t *params)
...
     }
 
-    /* Set calibration so that current register resolution is 0.01 mA */
-    uint32_t calib = 4096000 / params->rshunt_mohm;
+    /* Set calibration register based on the shunt resistor.
+     * ==> Current will be in mA
+     */
+    uint32_t calib = CURRENT_QUOTIENT / params->rshunt_mohm;
+    /* Change resultion from mA to 0.01 mA by multiplying with 100 */
+    calib *= 100;
```
Comment (benpicco) [inline drivers/ina2xx/ina2xx.c]: Possible loss of precision compared to the previous version?

File: drivers/ina2xx/ina2xx.c (hunk 3 above)
Comment (maribu) [reply drivers/ina2xx/ina2xx.c]: don

File: examples/default/Makefile (hunk 4 above)
Comment (maribu) [reply examples/default/Makefile]: Nice catch!

Comment (maribu) [reply examples/default/Makefile]: Fixed. Thanks for spotting this. That would have been embaressing again :-/

File: sys/auto_init/saul/auto_init_ina2xx.c (hunk 5 above)
Comment (maribu) [reply sys/auto_init/saul/auto_init_ina2xx.c]: Done

File: sys/auto_init/saul/auto_init_ina2xx.c (hunk 6 above)
Comment (maribu) [reply sys/auto_init/saul/auto_init_ina2xx.c]: Done

File: drivers/ina2xx/ina2xx.c (hunk 11 above)
Comment (maribu) [reply drivers/ina2xx/ina2xx.c]: Yes, indeed :-/ Fixed.

File: sys/auto_init/saul/auto_init_ina2xx.c (hunk 7 above)
Comment (maribu) [reply sys/auto_init/saul/auto_init_ina2xx.c]: Done

**Comment by benpicco** [comment]
Alright, you can squash!
I assume you ran some tests again too?

**Comment by maribu** [comment]
I just rebased (with no changes) on top of master, to make testing easier with the board currently connected to my PC.

**Comment by maribu** [comment]
<details><summary>Output of `tests/driver_ina2xx` on the ATmega1284P with no load on the attached INA219</summary>

```
2019-11-22 19:58:17,740 # main(): This is RIOT! (Version: 2020.01-devel-863-g3c4fe-ina2xx)
2019-11-22 19:58:17,779 # INA2XX sensor driver test application
2019-11-22 19:58:17,781 # 
2019-11-22 19:58:17,839 # Initializing INA2XX sensor at I2C_0, address 0x00000040
2019-11-22 19:58:17,844 # [OK]
2019-11-22 19:58:17,896 # +------------+--------------+----------+--------+
2019-11-22 19:58:17,951 # | U_Bus [mV] | U_Shunt [µV] |  I [µA]  | P [µW] |
2019-11-22 19:58:18,003 # +------------+--------------+----------+--------+
2019-11-22 19:58:18,055 # |       3328 |            0 |        0 |      0 |
2019-11-22 19:58:18,112 # |       3328 |            0 |        0 |      0 |
2019-11-22 19:58:18,259 # |       3328 |            0 |        0 |      0 |
2019-11-22 19:58:18,407 # |       3332 |            0 |        0 |      0 |
2019-11-22 19:58:18,555 # |       3332 |            0 |        0 |      0 |
2019-11-22 19:58:18,703 # |       3332 |            0 |        0 |      0 |
2019-11-22 19:58:18,850 # |       3332 |            0 |        0 |      0 |
2019-11-22 19:58:18,998 # |       3332 |            0 |        0 |      0 |
2019-11-22 19:58:19,146 # |       3332 |            0 |        0 |      0 |
2019-11-22 19:58:19,294 # |       3332 |            0 |        0 |      0 |
2019-11-22 19:58:19,441 # |       3332 |            0 |        0 |      0 |
2019-11-22 19:58:19,589 # |       3332 |            0 |        0 |      0 |
2019-11-22 19:58:19,737 # |       3332 |            0 |        0 |      0 |
2019-11-22 19:58:19,884 # |       3332 |            0 |        0 |      0 |
2019-11-22 19:58:20,032 # |       3332 |            0 |        0 |      0 |
```

</details>

<details><summary>Same, blue LED as load</summary>

```
2019-11-22 20:05:48,123 # main(): This is RIOT! (Version: 2020.01-devel-863-g3c4fe-ina2xx)
2019-11-22 20:05:48,162 # INA2XX sensor driver test application
2019-11-22 20:05:48,163 # 
2019-11-22 20:05:48,222 # Initializing INA2XX sensor at I2C_0, address 0x00000040
2019-11-22 20:05:48,227 # [OK]
2019-11-22 20:05:48,279 # +------------+--------------+----------+--------+
2019-11-22 20:05:48,334 # | U_Bus [mV] | U_Shunt [µV] |  I [µA]  | P [µW] |
2019-11-22 20:05:48,386 # +------------+--------------+----------+--------+
2019-11-22 20:05:48,437 # |       3320 |         2770 |     2780 |   9200 |
2019-11-22 20:05:48,495 # |       3320 |         2770 |     2780 |   9200 |
2019-11-22 20:05:48,642 # |       3320 |         2770 |     2780 |   9200 |
2019-11-22 20:05:48,790 # |       3320 |         2780 |     2780 |   9200 |
2019-11-22 20:05:48,938 # |       3320 |         2770 |     2780 |   9200 |
2019-11-22 20:05:49,086 # |       3320 |         2780 |     2780 |   9200 |
2019-11-22 20:05:49,234 # |       3320 |         2780 |     2780 |   9200 |
```

</details>

<details><summary>Same, green LED as load</summary>

```
2019-11-22 20:08:03,724 # main(): This is RIOT! (Version: 2020.01-devel-863-g3c4fe-ina2xx)
2019-11-22 20:08:03,763 # INA2XX sensor driver test application
2019-11-22 20:08:03,764 # 
2019-11-22 20:08:03,823 # Initializing INA2XX sensor at I2C_0, address 0x00000040
2019-11-22 20:08:03,828 # [OK]
2019-11-22 20:08:03,880 # +------------+--------------+----------+--------+
2019-11-22 20:08:03,935 # | U_Bus [mV] | U_Shunt [µV] |  I [µA]  | P [µW] |
2019-11-22 20:08:03,987 # +------------+--------------+----------+--------+
2019-11-22 20:08:04,039 # |       3320 |         2970 |     2980 |  10000 |
2019-11-22 20:08:04,096 # |       3320 |         2970 |     2980 |  10000 |
2019-11-22 20:08:04,243 # |       3320 |         2970 |     2980 |  10000 |
2019-11-22 20:08:04,391 # |       3320 |         2980 |     2980 |  10000 |
2019-11-22 20:08:04,539 # |       3316 |         2970 |     2980 |  10000 |
2019-11-22 20:08:04,687 # |       3320 |         2980 |     2980 |  10000 |
2019-11-22 20:08:04,835 # |       3320 |         2980 |     2980 |  10000 |
2019-11-22 20:08:04,982 # |       3320 |         2980 |     2980 |  10000 |
2019-11-22 20:08:05,130 # |       3320 |         2980 |     2980 |  10000 |
2019-11-22 20:08:05,278 # |       3320 |         2980 |     2980 |  10000 |
2019-11-22 20:08:05,426 # |       3320 |         2980 |     2980 |  10000 |
2019-11-22 20:08:05,574 # |       3320 |         2980 |     2980 |  10000 |
2019-11-22 20:08:05,721 # |       3320 |         2980 |     2980 |  10000 |
2019-11-22 20:08:05,869 # |       3320 |         2980 |     2980 |  10000 |
2019-11-22 20:08:06,017 # |       3320 |         2980 |     2980 |  10000 |
2019-11-22 20:08:06,165 # |       3320 |         2990 |     3000 |  10000 |
2019-11-22 20:08:06,313 # |       3320 |         2990 |     3000 |  10000 |
2019-11-22 20:08:06,460 # |       3320 |         2990 |     3000 |  10000 |
2019-11-22 20:08:06,608 # |       3320 |         2990 |     3000 |  10000 |
2019-11-22 20:08:06,756 # |       3320 |         2990 |     3000 |  10000 |
2019-11-22 20:08:06,903 # |       3320 |         2990 |     3000 |  10000 |
2019-11-22 20:08:07,051 # |       3320 |         2990 |     3000 |  10000 |
2019-11-22 20:08:07,199 # |       3320 |         2990 |     3000 |  10000 |
2019-11-22 20:08:07,347 # |       3320 |         2990 |     3000 |  10000 |
2019-11-22 20:08:07,495 # |       3320 |         2990 |     3000 |  10000 |
2019-11-22 20:08:07,643 # |       3320 |         2990 |     3000 |  10000 |
2019-11-22 20:08:07,790 # |       3320 |         2990 |     3000 |  10000 |
```

</details>


<details><summary>Same, blue **and** green LED as load (in parallel)</summary>

```2019-11-22 20:09:02,222 # main(): This is RIOT! (Version: 2020.01-devel-863-g3c4fe-ina2xx)
2019-11-22 20:09:02,262 # INA2XX sensor driver test application
2019-11-22 20:09:02,263 # 
2019-11-22 20:09:02,321 # Initializing INA2XX sensor at I2C_0, address 0x00000040
2019-11-22 20:09:02,326 # [OK]
2019-11-22 20:09:02,378 # +------------+--------------+----------+--------+
2019-11-22 20:09:02,433 # | U_Bus [mV] | U_Shunt [µV] |  I [µA]  | P [µW] |
2019-11-22 20:09:02,485 # +------------+--------------+----------+--------+
2019-11-22 20:09:02,537 # |       3308 |         5220 |     5220 |  17200 |
2019-11-22 20:09:02,594 # |       3308 |         5220 |     5220 |  17200 |
2019-11-22 20:09:02,742 # |       3308 |         5220 |     5220 |  17200 |
2019-11-22 20:09:02,890 # |       3308 |         5230 |     5240 |  17200 |
2019-11-22 20:09:03,038 # |       3308 |         5230 |     5240 |  17200 |
2019-11-22 20:09:03,186 # |       3308 |         5230 |     5240 |  17200 |
2019-11-22 20:09:03,333 # |       3308 |         5230 |     5240 |  17200 |
2019-11-22 20:09:03,481 # |       3308 |         5240 |     5240 |  17200 |
2019-11-22 20:09:03,629 # |       3308 |         5240 |     5240 |  17200 |
2019-11-22 20:09:03,777 # |       3308 |         5240 |     5240 |  17200 |
2019-11-22 20:09:03,925 # |       3308 |         5240 |     5240 |  17200 |
2019-11-22 20:09:04,072 # |       3308 |         5240 |     5240 |  17200 |
2019-11-22 20:09:04,220 # |       3308 |         5240 |     5240 |  17200 |
2019-11-22 20:09:04,368 # |       3308 |         5250 |     5260 |  17600 |
2019-11-22 20:09:04,516 # |       3308 |         5250 |     5260 |  17600 |
2019-11-22 20:09:04,664 # |       3308 |         5250 |     5260 |  17600 |
2019-11-22 20:09:04,812 # |       3308 |         5250 |     5260 |  17600 |
2019-11-22 20:09:04,960 # |       3308 |         5250 |     5260 |  17600 |
2019-11-22 20:09:05,107 # |       3308 |         5250 |     5260 |  17600 |
2019-11-22 20:09:05,255 # |       3308 |         5260 |     5260 |  17600 |
2019-11-22 20:09:05,403 # |       3308 |         5250 |     5260 |  17600 |
2019-11-22 20:09:05,551 # |       3308 |         5250 |     5260 |  17600 |
2019-11-22 20:09:05,699 # |       3308 |         5260 |     5260 |  17600 |
2019-11-22 20:09:05,846 # |       3308 |         5260 |     5260 |  17600 |
2019-11-22 20:09:05,994 # |       3308 |         5260 |     5260 |  17600 |
2019-11-22 20:09:06,142 # |       3308 |         5260 |     5260 |  17600 |
2019-11-22 20:09:06,290 # |       3308 |         5260 |     5260 |  17600 |
2019-11-22 20:09:06,438 # |       3308 |         5260 |     5260 |  17600 |
2019-11-22 20:09:06,585 # |       3308 |         5270 |     5280 |  17600 |
```

</details>

**Comment by benpicco** [comment]
Please squash again.

File: drivers/ina2xx/include/ina2xx_params.h [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+#define INA2XX_PARAMS               { .i2c         = INA2XX_PARAM_I2C,  \
+                                      .addr        = INA2XX_PARAM_ADDR, \
+                                      .config      = INA2XX_PARAM_CONFIG, \
+                                      .rshunt_mohm = INA2XX_PARAM_RSHUNT_MOHM, \
+                                      .i_range     = INA2XX_PARAM_I_RANGE }
+#endif
+/**@}*/
+
+/**
+ * @name    Set defaul SAUL info for the INA2XX
```
Comment (benpicco) [inline drivers/ina2xx/include/ina2xx_params.h]:
```suggestion
 * @name    Set default SAUL info for the INA2XX
```

Comment (maribu) [reply drivers/ina2xx/include/ina2xx_params.h]: Done. Nice catch!

**Comment by maribu** [comment]
@benpicco, @gschorcht, @smlng: Thanks for your comments!

**Pull request merged**